src/main/java/net/neoforged/neoforge/event/ItemAttributeModifierEvent.java

```diff
@@ -5,118 +5,231 @@
-     * @param modifier  Modifier instance
-     * @return True if an attribute was removed, false if no change
+     * @return true if any modifiers were removed
      */
-    public boolean removeModifier(Holder<Attribute> attribute, AttributeModifier modifier) {
-        return getModifiableMap().remove(attribute, modifier);
+    public boolean removeIf(Predicate<ItemAttributeModifiers.Entry> condition) {
+        return getBuilder().removeIf(condition);
     }
 
     /**
-     * Removes all modifiers for the given attribute
+     * Removes all modifiers for the given attribute.
      * 
-     * @param attribute Attribute
-     * @return Collection of removed modifiers
+     * @return true if any modifiers were removed
      */
-    public Collection<AttributeModifier> removeAttribute(Holder<Attribute> attribute) {
-        return getModifiableMap().removeAll(attribute);
+    public boolean removeAllModifiersFor(Holder<Attribute> attribute) {
+        return getBuilder().removeIf(entry -> entry.attribute().equals(attribute));
     }
 
     /**
-     * Removes all modifiers for all attributes
+     * Removes all modifiers for all attributes.
      */
     public void clearModifiers() {
-        getModifiableMap().clear();
+        getBuilder().clear();
     }
 
-    /** Gets the slot containing this stack */
-    public EquipmentSlot getSlotType() {
-        return this.slotType;
+    /**
+     * Builds a new {@link ItemAttributeModifiers} with the results of this event, returning the
+     * {@linkplain #getDefaultModifiers() default modifiers} if no changes were made.
+     */
+    public ItemAttributeModifiers build() {
+        return this.builder == null ? this.defaultModifiers : this.builder.build(this.defaultModifiers.showInTooltip());
     }
 
-    /** Gets the item stack instance */
-    public ItemStack getItemStack() {
-        return this.stack;
+    /**
+     * Returns the builder used for adjusting the attribute modifiers, creating it if it does not already exist.
+     */
+    private ItemAttributeModifiersBuilder getBuilder() {
+        if (this.builder == null) {
+            this.builder = new ItemAttributeModifiersBuilder(this.defaultModifiers);
+        }
+
+        return this.builder;
+    }
+
+    /**
+     * Advanced version of {@link ItemAttributeModifiers.Builder} which supports removal and better sanity-checking.
+     * <p>
+     * The original builder only supports additions and does not guarantee that no duplicate modifiers exist for a given id.
+     */
+    private static class ItemAttributeModifiersBuilder {
+        private List<ItemAttributeModifiers.Entry> entries;
+        private Map<Key, ItemAttributeModifiers.Entry> entriesByKey;
+
+        ItemAttributeModifiersBuilder(ItemAttributeModifiers defaultModifiers) {
+            this.entries = new LinkedList<>();
+            this.entriesByKey = new HashMap<>(defaultModifiers.modifiers().size());
+
+            defaultModifiers.modifiers().forEach(entry -> {
```

Comment:
We should use a for loop rather than a `forEach` here to avoid the capturing lambda